server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/schema/NormalSchemaFetcher.java

```diff
@@ -47,30 +51,104 @@ class NormalSchemaFetcher {
+    }
+    return patternTree;
+  }
+
+  private PathPatternTree computePatternTreeNeededReFetch(
+      List<PartialPath> devicePathList,
+      List<String[]> measurementsList,
+      List<Integer> indexOfTargetDevices,
+      List<List<Integer>> indexOfTargetMeasurementsList) {
+    PathPatternTree patternTree = new PathPatternTree();
+    int deviceIndex;
+    for (int i = 0, size = indexOfTargetDevices.size(); i < size; i++) {
+      deviceIndex = indexOfTargetDevices.get(i);
+      for (int measurementIndex : indexOfTargetMeasurementsList.get(i)) {
+        patternTree.appendFullPath(
+            devicePathList.get(deviceIndex), measurementsList.get(deviceIndex)[measurementIndex]);
+      }
+    }
+    return patternTree;
+  }
+
   List<Integer> processNormalTimeSeries(
       ISchemaComputationWithAutoCreation schemaComputationWithAutoCreation) {
+    // [Step 1] Cache 1. compute measurements and record logical views.
     List<Integer> indexOfMissingMeasurements =
         schemaCache.computeWithoutTemplate(schemaComputationWithAutoCreation);
+    // [Step 2] Cache 2. process recorded logical views.
+    Pair<List<Integer>, List<String>> missedIndexAndPathString =
+        schemaCache.computeLogicalViewWithoutTemplate(schemaComputationWithAutoCreation);
+    List<Integer> indexOfMissingLogicalView = missedIndexAndPathString.left;
+    List<String> missedPathStringOfLogicalView = missedIndexAndPathString.right;
     // all schema can be taken from cache
-    if (indexOfMissingMeasurements.isEmpty()) {
+    if (indexOfMissingMeasurements.isEmpty() && indexOfMissingLogicalView.isEmpty()) {
       return indexOfMissingMeasurements;
     }
-
-    // try fetch the missing schema from remote and cache fetched schema
-    ClusterSchemaTree remoteSchemaTree =
-        clusterSchemaFetchExecutor.fetchSchemaOfOneDevice(
-            schemaComputationWithAutoCreation.getDevicePath(),
-            schemaComputationWithAutoCreation.getMeasurements(),
-            indexOfMissingMeasurements);
+    // [Step 3] Fetch 1. fetch schema from remote. Process logical view first; then process
+    // measurements.
+    // try fetch the missing raw schema from remote and cache fetched schema
+    ClusterSchemaTree remoteSchemaTree;
+    if (missedPathStringOfLogicalView.isEmpty()) {
+      remoteSchemaTree =
+          clusterSchemaFetchExecutor.fetchSchemaOfOneDevice(
+              schemaComputationWithAutoCreation.getDevicePath(),
+              schemaComputationWithAutoCreation.getMeasurements(),
+              indexOfMissingMeasurements);
+    } else {
+      PathPatternTree patternTree =
+          computePatternTreeNeededReFetch(
+              schemaComputationWithAutoCreation.getDevicePath(),
+              schemaComputationWithAutoCreation.getMeasurements(),
+              indexOfMissingMeasurements);
+      remoteSchemaTree =
+          fetchWithPatternTreeAndFullPaths(patternTree, missedPathStringOfLogicalView);
```

Comment:
Separate the pattern tree construction and fetch operation.

Reply:
fixed